syft/format/cyclonedxxml/encoder_test.go

```diff
@@ -0,0 +1,112 @@
+	testutil.AssertEncoderAgainstGoldenSnapshot(t,
+		testutil.EncoderSnapshotTestConfig{
+			Subject:                     testutil.DirectoryInput(t, dir),
+			Format:                      DefaultFormatEncoder(),
+			UpdateSnapshot:              *updateSnapshot,
+			PersistRedactionsInSnapshot: true,
+			IsJSON:                      false,
+			Redactor:                    redactor(dir),
+		},
+	)
+}
+
+func TestCycloneDxImageEncoder(t *testing.T) {
+	testImage := "image-simple"
+	testutil.AssertEncoderAgainstGoldenImageSnapshot(t,
+		testutil.ImageSnapshotTestConfig{
+			Image:               testImage,
+			UpdateImageSnapshot: *updateImage,
+		},
+		testutil.EncoderSnapshotTestConfig{
+			Subject:                     testutil.ImageInput(t, testImage),
+			Format:                      DefaultFormatEncoder(),
+			UpdateSnapshot:              *updateSnapshot,
+			PersistRedactionsInSnapshot: true,
+			IsJSON:                      false,
+			Redactor:                    redactor(),
+		},
+	)
+}
+
+func redactor(values ...string) testutil.Redactor {
+	return testutil.NewRedactions().
+		WithValuesRedacted(values...).
+		WithPatternRedactors(
+			map[string]string{
+				// serial numbers
+				`serialNumber="[a-zA-Z0-9\-:]+`: `serialNumber="redacted`,
+
+				// dates
+				`([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])[Tt]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\.[0-9]+)?(([Zz])|([+|\-]([01][0-9]|2[0-3]):[0-5][0-9]))`: `redacted`,
+
+				// image hashes
+				`sha256:[A-Za-z0-9]{64}`: `sha256:redacted`,
+
+				// BOM refs
+				`bom-ref="[a-zA-Z0-9\-:]+"`: `bom-ref:redacted`,
+			},
+		)
+}
+
+func TestSupportedVersions(t *testing.T) {
+	encs := DefaultFormatEncoders()
+	require.NotEmpty(t, encs)
+
+	versions := SupportedVersions()
+	require.Equal(t, len(versions), len(encs))
+
+	subject := testutil.DirectoryInput(t, t.TempDir())
+	dec := NewFormatDecoder()
+
+	for _, enc := range encs {
+		t.Run(enc.Version(), func(t *testing.T) {
+			require.Contains(t, versions, enc.Version())
+
+			var buf bytes.Buffer
+			require.NoError(t, enc.Encode(&buf, subject))
+
+			id, version := dec.Identify(buf.Bytes())
+			require.Equal(t, enc.ID(), id)
+			require.Equal(t, enc.Version(), version)
+
+			var s *sbom.SBOM
+			var err error
+			s, id, version, err = dec.Decode(buf.Bytes())
+			require.NoError(t, err)
+			require.Equal(t, enc.ID(), id)
+			require.Equal(t, enc.Version(), version)
+
+			assert.Equal(t, len(subject.Relationships), len(s.Relationships), "mismatched relationship count")
+
+			if !assert.Equal(t, subject.Artifacts.Packages.PackageCount(), s.Artifacts.Packages.PackageCount(), "mismatched package count") {
+				t.Logf("expected: %d", subject.Artifacts.Packages.PackageCount())
+				for _, p := range subject.Artifacts.Packages.Sorted() {
+					t.Logf("  - %s", p.String())
+				}
+				t.Logf("actual: %d", s.Artifacts.Packages.PackageCount())
+				for _, p := range s.Artifacts.Packages.Sorted() {
+					t.Logf("  - %s", p.String())
+				}
+			}
```

Comment:
Actually, this is a new form of test for each format that did not exist previously. The point of this test isn't to test in depth the correctness of the SBOM content (this is what other unit tests for each format are for) instead this is mean to ensure that there is some parity with encode-identify-decode cycles, sanity checking the high level results (did we get the expected format id and version for the objects used). The package check is a bonus, ensuring that the results when going through this cycle didn't get cleared/dropped.

Reply:
The suggestion wasn't to improve the accuracy of the test but rather to improve the  error case if the test fails. As is, I'll get something like:

```
mismatched package count
expected: 3
  - some-package-1
  - some-package-2
  - some-package-3
actual: 2
  - some-package-1
  - some-package-3
```
now I have to manually examine the lists or use some scripting to find the diff, whereas if we just had a couple lists of names:
```
expected := packageNames(subject.Artifacts.Packages.Sorted()...)
got := packageNames(s.Artifacts.Packages.Sorted()...)
require.Equal(t, expected, got)
```
... we'd get something that shows us explicitly what's wrong so it's easier to dive in and figure out why a specific thing did or didn't show up unexpectedly.

Reply:
Given the test input, I think there will only ever be 2 to 3 packages (it would be really close to the output in your example). Also , these are count based, and not all formats will produce the same count (spdx). There are baked offsets for these cases. I'd need to keep the name sets in those cases.

If the input was several packages, I'd strongly agree and make the update, but since it's always very few packages I don't think the output would be that difficult to read.